### PR TITLE
Allow SMask be resumed after restore() and better transform after SMask

### PR DESCRIPTION
**Pull request by yurydelendik**
... also fixes finalizing SMask at the end of operator list.

Fixes most of #7011

### CODE REVIEW

**Comment by yurydelendik** [comment]
TODO:
- [ ] include bug852992_reduced.pdf as a test (Bug 852992)
- [ ] add https://www.sec.gov/investor/pubs/sec-guide-to-proxy-brochures.pdf test (Issue 6165)
- [ ] add https://bugzilla.mozilla.org/attachment.cgi?id=8653724 test (Bug 1199237)

#7143 is combination of (Issue 6165 and Bug 1199237 but in weird way)

File: src/display/canvas.js [hunk 1]
```diff
@@ -869,6 +870,11 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
     },
 
     endDrawing: function CanvasGraphics_endDrawing() {
+      // Finishing all opened operations such as SMask group painting.
+      if (this.current.activeSMask !== null) {
+          this.endSMaskGroup();
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: Nit: please reduce the indentation here.

File: src/display/canvas.js [hunk 2]
```diff
@@ -1016,6 +1033,43 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
       this.groupStack.push(currentCtx);
       this.groupLevel++;
     },
+    suspendSMaskGroup: function CanvasGraphics_endSMaskGroup() {
+      // Similar to endSMaskGroup, the intermidiate canvas has to be composed
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: Nit: incorrect spelling, should be `intermediate`.

File: src/display/canvas.js [hunk 3]
```diff
@@ -1016,6 +1033,43 @@ var CanvasGraphics = (function CanvasGraphicsClosure() {
...
+      var groupCtx = this.ctx;
+      this.groupLevel--;
+      this.ctx = this.groupStack.pop();
+
+      composeSMask(this.ctx, this.current.activeSMask, groupCtx);
+      this.ctx.restore();
+      this.ctx.save(); // save is needed since SMask will be resumed.
+      copyCtxState(groupCtx, this.ctx);
+
+      // Saving save for resuming.
```
Comment (Snuffleupagus) [inline src/display/canvas.js]: Nit: Shouldn't this be `Saving state ...` instead?

**Comment by Snuffleupagus** [comment]
This looks great, awesome work! And the added comments was most helpful when trying to make sense of the code.
r=me, with the final nit addressed, and passing tests.

**Comment by Snuffleupagus** [comment]
For future reference, this also fixed https://bugzilla.mozilla.org/show_bug.cgi?id=1025247.

**Pull request merged**